File: node-bindings/classic-node/__test__/installation_paths_conformance.ts
import { chmod, mkdir, mkdtemp, readFile, readdir, rm, stat, writeFile } from "node:fs/promises";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { DocsPathFinder, DocumentsChecker, GamePathFinder } from "../index.js";
import * as classic from "../index.js";

type JsonObject = Record<string, any>;

/** Read a complete sorted durable tree, including unexpected native writes. */
async function inventory(root: string, prefix = ""): Promise<{files: JsonObject[], directories: string[]}> {
  const files: JsonObject[] = [];
  const directories: string[] = [];
  for (const entry of await readdir(join(root, prefix), { withFileTypes: true })) {
    const path = prefix ? `${prefix}/${entry.name}` : entry.name;
    if (entry.isDirectory()) {
      directories.push(path);
      const nested = await inventory(root, path);
      files.push(...nested.files); directories.push(...nested.directories);
    } else if (entry.isFile()) files.push({path, content: await readFile(join(root, path), "utf8")});
    else throw new Error("unexpected installation artifact");
  }
  files.sort((a, b) => a.path < b.path ? -1 : a.path > b.path ? 1 : 0);
  return {files, directories: directories.sort()};
}

/** Execute each native public operation on validated owned caches before any OS fallback. */
export async function observeInstallationPaths(fixture: JsonObject): Promise<JsonObject> {
  const game = fixture.gamePath, docs = fixture.docsPath;
  if (!((game === "game" && docs === "docs") || (game === "Game Folder" && docs === "Docs Folder"))) {
    throw new Error("unsupported installation cache paths");
  }
  const root = await mkdtemp(join(tmpdir(), "classic-installation-conformance-"));
  const previous = process.cwd();
  try {
    await mkdir(join(root, game)); await mkdir(join(root, docs));
    await writeFile(join(root, "CLASSIC Main.yaml"), fixture.registryYaml, "utf8");
    if (Object.keys(fixture.files).join() !== `${game}/Fallout4.exe` || fixture.files[`${game}/Fallout4.exe`] !== "owned executable marker") {
      throw new Error("installation fixture needs a valid cached executable");
    }
    await writeFile(join(root, game, "Fallout4.exe"), "owned executable marker");
    // The dedicated receipt process is serial; relative paths exclude host
    // parent names such as OneDrive from the document-checking contract.
    process.chdir(root);
    const gameFinder = new GamePathFinder("Fallout4.exe", null, "Fallout4", false);
    const docsFinder = new DocsPathFinder("My Games/Fallout4");
    docsFinder.setSteamAppId(12345);
    gameFinder.validateGamePath(game); docsFinder.validateDocsPath(docs);
    const checker = new DocumentsChecker("Fallout4");
    const checks = checker.runAllChecks(docs);
    if (checker.checkOnedriveInPath(docs) !== null) throw new Error("owned path unexpectedly reports OneDrive");
    ["Fallout4.ini", "Fallout4Custom.ini", "Fallout4Prefs.ini"].forEach((name, index) => {
      const check = checker.validateIniFile(docs, name);
      if (check.exists || check.isValid || !check.hasIssue || check.iniName !== name || check.message !== checks[index]) throw new Error("INI diagnosis disagrees with aggregate check");
    });
    docsFinder.validateIniFiles(docs, []);
    let missingRejected = false;
    try { docsFinder.validateIniFiles(docs, ["Fallout4.ini"]); } catch (error) { if (!(error instanceof Error) || !error.message.includes("Fallout4.ini")) throw error; missingRejected = true; }
    if (!missingRejected) throw new Error("missing required INI was accepted");
    await mkdir("validation/owned/scan", {recursive: true});
    classic.validateCustomScanPath("validation/owned/scan");
    classic.validateSettingsPath(game, "Game Path", ["Fallout4.exe"]);
    classic.validateSettingsPaths(game, docs, "validation/owned/scan", "Fallout4.exe");
    classic.checkDriveExists(root); classic.checkReadPermissions(game); classic.checkWritePermissions(game);
    classic.validatePathWithPermissions(game, true, true);
    if (!classic.isValidPath(game) || classic.isValidPath("missing-path")) throw new Error("path existence alias disagrees with owned tree");
    classic.validateRequiredFiles(game, ["Fallout4.exe"]);
    let requiredRejected = false;
    try { classic.validateRequiredFiles(game, ["missing.ini"]); } catch (error) {
      if (!(error instanceof Error) || !error.message.includes("missing.ini")) throw error;
      requiredRejected = true;
    }
    if (!requiredRejected) throw new Error("required-files alias accepted an absent file");
    const readonlyFile = "validation/owned/scan/readonly.txt";
    await writeFile(readonlyFile, "retained bytes");
    await chmod(readonlyFile, 0o444);
    if (((await stat(readonlyFile)).mode & 0o200) !== 0) throw new Error("readonly precondition was not established");
    classic.removeReadonly(readonlyFile);
    if (((await stat(readonlyFile)).mode & 0o200) === 0 || await readFile(readonlyFile, "utf8") !== "retained bytes") throw new Error("readonly removal changed bytes or failed to restore write access");
    if (classic.isRestrictedPath("validation/owned/scan") || !classic.isRestrictedPath("Windows/System32/test")) throw new Error("restricted-path classification changed");
    if (!classic.isValidExecutablePath(`${game}/Fallout4.exe`) || classic.isValidExecutablePath("CLASSIC Main.yaml")) throw new Error("executable path classification changed");
    await rm("validation", {recursive: true});
    await writeFile("path-detection.log", `plugin directory = "${game}/Data/F4SE/Plugins"\n`);
    if (classic.parseXseLog("path-detection.log") !== game) throw new Error("XSE log path extraction changed game root");
    await rm("path-detection.log");
    return {gamePath: gameFinder.findGamePath(game, null), docsPath: docsFinder.findDocsPath(docs),
      checks, ...await inventory(root)};
  } finally {
    process.chdir(previous);
    await rm(root, { recursive: true, force: true });
  }
}

File: python-bindings/tests/installation_paths_conformance.py
"""Observe cached paths and missing INI reports against an invocation-owned tree."""

import os
import stat
import tempfile
from pathlib import Path


def observe_installation_paths(fixture):
    """Execute native public methods with valid caches, restoring cwd before cleanup."""
    import classic_path

    game, docs = fixture["gamePath"], fixture["docsPath"]
    if (game, docs) not in {("game", "docs"), ("Game Folder", "Docs Folder")}:
        raise ValueError("unsupported installation cache paths")
    with tempfile.TemporaryDirectory(
        prefix="classic-installation-conformance-"
    ) as directory:
        root = Path(directory)
        (root / game).mkdir()
        (root / docs).mkdir()
        (root / "CLASSIC Main.yaml").write_bytes(
            fixture["registryYaml"].encode("utf-8")
        )
        if fixture["files"] != {game + "/Fallout4.exe": "owned executable marker"}:
            raise ValueError("installation fixture needs a valid cached executable")
        (root / game / "Fallout4.exe").write_bytes(b"owned executable marker")
        previous = Path.cwd()
        try:
            # Relative inputs keep host parent names (for example OneDrive) out
            # of the public path checker. Receipt processes execute serially.
            os.chdir(root)
            game_finder = classic_path.GamePathFinder(
                "Fallout4.exe", None, "Fallout4", False
            )
            docs_finder = classic_path.DocsPathFinder("My Games/Fallout4")
            docs_finder.set_steam_app_id(12345)
            game_finder.validate_game_path(game)
            docs_finder.validate_docs_path(docs)
            result = {
                "gamePath": game_finder.find_game_path(
                    cached_path=game, xse_log_path=None
                ),
                "docsPath": docs_finder.find_docs_path(cached_path=docs),
                "checks": classic_path.DocumentsChecker("Fallout4").run_all_checks(
                    docs
                ),
            }
            checker = classic_path.DocumentsChecker("Fallout4")
            if checker.check_onedrive_in_path(docs) is not None:
                raise ValueError("owned documents path unexpectedly reports OneDrive")
            for index, name in enumerate(
                ("Fallout4.ini", "Fallout4Custom.ini", "Fallout4Prefs.ini")
            ):
                check = checker.validate_ini_file(docs, name)
                if (
                    check.exists
                    or check.is_valid
                    or not check.has_issue()
                    or check.ini_name != name
                    or check.message != result["checks"][index]
                ):
                    raise ValueError(
                        "per-file INI diagnosis disagrees with aggregate check"
                    )
            docs_finder.validate_ini_files(docs, [])
            try:
                docs_finder.validate_ini_files(docs, ["Fallout4.ini"])
            except (FileNotFoundError, ValueError) as failure:
                if "Fallout4.ini" not in str(failure):
                    raise
            else:
                raise ValueError("missing required INI was accepted")
            # Keep lexical restriction checks independent of host Temp folders
            # (which commonly include AppData) using an owned relative subtree.
            scan = Path("validation/owned/scan")
            scan.mkdir(parents=True)
            validator = classic_path.PathValidator
            validator.validate_custom_scan_path(str(scan))
            validator.validate_settings_path(game, "Game Path", ["Fallout4.exe"])
            validator.validate_settings_paths(game, docs, str(scan), "Fallout4.exe")
            validator.check_drive_exists(str(root))
            validator.check_read_permissions(game)
            validator.check_write_permissions(game)
            validator.validate_path_with_permissions(game, True, True)
            if not validator.is_valid_path(game) or validator.is_valid_path(
                "missing-path"
            ):
                raise ValueError("path existence alias disagrees with owned tree")
            validator.validate_required_files(game, ["Fallout4.exe"])
            try:
                validator.validate_required_files(game, ["missing.ini"])
            except FileNotFoundError as failure:
                if "missing.ini" not in str(failure):
                    raise
            else:
                raise ValueError("required-files alias accepted an absent file")
            readonly_file = scan / "readonly.txt"
            readonly_file.write_bytes(b"retained bytes")
            readonly_file.chmod(stat.S_IREAD)
            if readonly_file.stat().st_mode & stat.S_IWRITE:
                raise ValueError("readonly precondition was not established")
            classic_path.remove_readonly(str(readonly_file))
            if (
                not readonly_file.stat().st_mode & stat.S_IWRITE
                or readonly_file.read_bytes() != b"retained bytes"
            ):
                raise ValueError(
                    "readonly removal changed bytes or failed to restore write access"
                )
            readonly_file.unlink()
            if validator.is_restricted_path(
                str(scan)
            ) or not validator.is_restricted_path("Windows/System32/test"):
                raise ValueError(
                    "restricted-path classification disagrees with owned input"
                )
            if not validator.is_valid_executable_path(
                game + "/Fallout4.exe"
            ) or validator.is_valid_executable_path("CLASSIC Main.yaml"):
                raise ValueError(
                    "executable path classification lost file extension semantics"
                )
            scan.rmdir()
            scan.parent.rmdir()
            scan.parent.parent.rmdir()
            log_path = Path("path-detection.log")
            log_path.write_text(
                f'plugin directory = "{game}/Data/F4SE/Plugins"\n', encoding="utf-8"
            )
            if classic_path.GamePathFinder.parse_xse_log(str(log_path)) != game:
                raise ValueError("XSE log path extraction did not preserve game root")
            log_path.unlink()
            result["files"] = [
                {
                    "path": path.relative_to(root).as_posix(),
                    "content": path.read_bytes().decode("utf-8"),
                }
                for path in sorted(root.rglob("*"))
                if path.is_file()
            ]
            result["directories"] = sorted(
                path.relative_to(root).as_posix()
                for path in root.rglob("*")
                if path.is_dir()
            )
            return result
        finally:
            os.chdir(previous)
